tilings/algorithms/factor.py

```diff
@@ -0,0 +1,280 @@
+
+        TODO: Describe the meaning or workable.
+        """
+        if not self.factorable():
+            return
+        assert isinstance(workable, bool)
+        return Rule(formal_step,
+                    factors,
+                    inferable=[False for _ in factors],
+                    workable=[workable for _ in factors],
+                    possibly_empty=[False for _ in factors],
+                    ignore_parent=workable,
+                    constructor=self.constructor)
+
+    def rule(self, workable=True):
+        return self._rule(self.factors(), self.formal_step(), workable)
+
+    def all_union_rules(self, workable=True):
+        """
+        Iterator over the rule for all possible union of factors.
+
+        A rule is yielded for reach reducible_factorisations
+
+        This generator includes a normal factor rule for the irreducible
+        factorisation.
+        """
+        for factorisation in self.reducible_factorisations():
+            yield self._rule(
+                factorisation,
+                self.formal_step(union=True),
+                workable=workable
+            )
+
+
+class FactorWithMonotoneInterleaving(Factor):
+    """
+    Algorithm to compute the factorisation with monotone interleaving of a
+    tiling.
+
+    Two active cells are in the same factor if they share an obstruction or
+    a requirement or if they are on the same row or column and are both
+    non-monotone.
+    """
+
+    def _unite_rows_and_cols(self):
+        """
+        Unite all active cell that are on the same row or column if both of
+        them are not monotone.
+
+        Override `Factor._unite_rows_and_cols`.
+        """
+        cell_pair_to_unite = (c for c in combinations(self._active_cells, r=2)
+                              if (self._same_row_or_col(c[0], c[1]) and
+                                  not self._tiling.is_monotone_cell(c[0]) and
+                                  not self._tiling.is_monotone_cell(c[1])))
+        for c1, c2 in cell_pair_to_unite:
+            self._unite_cells((c1, c2))
+
+    def formal_step(self, union=False):
+        """
+        Return a string that describe the operation performed on the tiling.
+        """
+        union_str = 'unions of ' if union else ''
+        return ("The {}factors with monotone interleaving of the "
+                "tiling.".format(union_str))
+
+    @property
+    def constructor(self):
+        """
+        Returns the type of constructor for the factorisation
+        """
+        return 'other'
+
+
+class FactorWithInterleaving(Factor):
+    """
+    Algorithm to compute the factorisation with interleaving of a tiling.
+
+    Two non-empty cells are in the same factor if they share an obstruction or
+    a requirement.
+    """
+
+    def _unite_rows_and_cols(self):
+        """
+        Override the `Factor._unite_rows_and_cols` to do nothing since
+        interleaving is allowed on row and column.
+        """
+        pass
```

Comment:
Perhaps instead overwrite the _unite_all function?

Reply:
I think we should let this one this way since the behaviour of interleaving factor is to not union any row and columns. If don't do anything the behaviour from the parent class will apply. I'll will also overide the _unite_all method.